### PR TITLE
[Refactor] Decoupling missingUpdateEventsTranscoder

### PR DESCRIPTION
**Pull request by KaterinaWire**
## What's new in this PR?

Move the logic related to Push notifications to the `NotificationStreamSync` class.

### CODE REVIEW

File: Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+    
+    public var zm_hasLastNotificationID: Bool {
+        return zm_lastNotificationID != nil
+    }
+}
+
+extension UUID {
+    func compare(withType1 uuid: UUID) -> ComparisonResult {
```
Comment (typfel) [inline Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift:56]: Should move to `wire-ios-utilities`https://github.com/wireapp/wire-ios-utilities/blob/develop/Source/Public/NSUUID+Type1.swift

But we can also do it a later stage.

File: Sources/Helpers/NotificationStreamSync.swift [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            for conversation in conversations {
+                if timestamp == nil {
+                    // In case we did not receive a payload we will add 1/10th to the last modified date of
+                    // the conversation to make sure it appears below the last message
+                    timestamp = conversation.lastModifiedDate?.addingTimeInterval(offset) ?? Date()
+                }
+                if let timestamp = timestamp {
+                    conversation.appendNewPotentialGapSystemMessage(users: conversation.localParticipants, timestamp: timestamp)
+                }
+            }
```
Comment (typfel) [inline Sources/Helpers/NotificationStreamSync.swift:163]: I would move this into a static method on `ZMConversation` in data model.

File: Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift (hunk 1 above)
Comment (KaterinaWire) [reply Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift:56]: Done

File: Sources/Helpers/DataModel/NSManagedObjectContext+ServerTimeDelta.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public extension NSManagedObjectContext {
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Sources/Helpers/DataModel/NSManagedObjectContext+ServerTimeDelta.swift]: Done

File: Sources/Helpers/NotificationStreamSync.swift [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+        return request
+    }
+    
+    private var lastUpdateEventID: UUID? {
+        return self.managedObjectContext.zm_lastNotificationID
+    }
+    
+    @objc(nextUUIDFromResponse:forListPaginator:)
+    public func nextUUID(from response: ZMTransportResponse!, forListPaginator paginator: ZMSimpleListRequestPaginator!) -> UUID! {
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Sources/Helpers/NotificationStreamSync.swift]: Discussed in 1:1 and came up with a decision to do this refactoring later.

File: Sources/Helpers/NotificationStreamSync.swift [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    func failedFetchingEvents()
+}
+
+public class NotificationStreamSync: NSObject, ZMRequestGenerator, ZMSimpleListRequestPaginatorSync {
+    
+    private var paginator: ZMSimpleListRequestPaginator?
+    private var notificationsTracker: NotificationsTracker?
+    private var listPaginator: ZMSimpleListRequestPaginator!
+    private var managedObjectContext: NSManagedObjectContext!
+    private var previouslyReceivedEventIDsCollection: PreviouslyReceivedEventIDsCollection?
```
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Sources/Helpers/NotificationStreamSync.swift]: I moved it to the strategy.

File: Sources/Helpers/NotificationStreamSync.swift (hunk 2 above)
Comment (KaterinaWire) [reply Sources/Helpers/NotificationStreamSync.swift]: Done

File: Resources/Configurations/version.xcconfig [hunk 6]
```diff
@@ -1,2 +1,2 @@
-MAJOR_VERSION = 200
-CURRENT_PROJECT_VERSION = 200.0.0
+MAJOR_VERSION = 201
+CURRENT_PROJECT_VERSION = 201.0.0
```
Comment (typfel) [inline Resources/Configurations/version.xcconfig:2]: I think you need to update the base branch that your are merging into.

File: Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift (hunk 1 above)
Comment (typfel) [reply Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift:56]: Still need to delete the code here though

File: Sources/Helpers/DataModel/NSManagedObjectContext+ServerTimeDelta.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (KaterinaWire) [reply Sources/Helpers/DataModel/NSManagedObjectContext+ServerTimeDelta.swift]: Removed

File: Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift (hunk 1 above)
Comment (KaterinaWire) [reply Sources/Helpers/DataModel/NSManagedObjectContext+LastNotificationID.swift]: Deleted

File: Resources/Configurations/version.xcconfig (hunk 6 above)
Comment (KaterinaWire) [reply Resources/Configurations/version.xcconfig]: Updated

File: Sources/Helpers/Notifications/PushNotificationStatus.swift [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        switch self {
+        case .done: return "done"
+        case .inProgress: return "inProgress"
+        }
+    }
+
+}
+
+@objcMembers
+open class PushNotificationStatus: NSObject, BackgroundNotificationFetchStatusProvider {
```
Comment (typfel) [inline Sources/Helpers/Notifications/PushNotificationStatus.swift:37]: We should also move the PushNotificationStatus tests

File: Sources/Helpers/ZMSimpleListRequestPaginator.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+@import Foundation;
+@import WireTransport;
+#import "ZMSingleRequestSync.h"
+#import "ZMRequestGenerator.h"
+
+@protocol ZMSimpleListRequestPaginatorSync;
+
+
+@interface ZMSimpleListRequestPaginator : NSObject <ZMRequestGenerator>
```
Comment (typfel) [inline Sources/Helpers/ZMSimpleListRequestPaginator.h:27]: We should also move the ZMSimpleListRequestPaginator tests

File: Sources/Helpers/Notifications/NotificationsTracker.swift [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+
+@objcMembers public class NotificationsTracker: NSObject {
```
Comment (typfel) [inline Sources/Helpers/Notifications/NotificationsTracker.swift:22]: We should also move the NotificationsTracker tests

File: Sources/Helpers/Notifications/PushNotificationStatus.swift (hunk 7 above)
Comment (KaterinaWire) [reply Sources/Helpers/Notifications/PushNotificationStatus.swift:37]: Done

File: Sources/Helpers/Notifications/NotificationsTracker.swift (hunk 9 above)
Comment (KaterinaWire) [reply Sources/Helpers/Notifications/NotificationsTracker.swift:22]: Done

File: Sources/Helpers/ZMSimpleListRequestPaginator.h (hunk 8 above)
Comment (typfel) [reply Sources/Helpers/ZMSimpleListRequestPaginator.h:27]: Also done it looks like?

File: Tests/Sources/MessagingTest.h [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+
+//@import WireTesting;
```
Comment (typfel) [inline Tests/Sources/MessagingTest.h:20]: Commented out code?

File: Tests/Sources/Utility/AnalyticsTests.swift [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    
+    override func tearDown() {
+        sharedContainerURL = nil
+        directory = nil
+        accountID = nil
+        analytics = nil
+        StorageStack.reset()
+        super.tearDown()
+    }
+    // FIX IT
```
Comment (typfel) [inline Tests/Sources/Utility/AnalyticsTests.swift:58]: Still to be fixed?

Comment (KaterinaWire) [reply Tests/Sources/Utility/AnalyticsTests.swift]: Fixed

File: Tests/Sources/MessagingTest.h (hunk 10 above)
Comment (KaterinaWire) [reply Tests/Sources/MessagingTest.h]: Fixed

File: Sources/Helpers/ZMSimpleListRequestPaginator.h (hunk 8 above)
Comment (KaterinaWire) [reply Sources/Helpers/ZMSimpleListRequestPaginator.h:27]: Done

**Comment by typfel** [comment]
Why did the `SyncStatus` need be moved to here?

**Comment by KaterinaWire** [comment]
> Why did the `SyncStatus` need be moved to here?

@typfel  I don't need it now. I reverted the changes.

**Pull request merged**